pkg/controllers/cluster/kubeconfig_controller.go

```diff
@@ -0,0 +1,308 @@
+			return ctrl.Result{}, err
+		}
+
+		var clusterCfg *clientcmdapi.Cluster
+		for _, v := range kubeCfg.Clusters {
+			clusterCfg = v
+			break
+		}
+
+		kubeconfig.Spec.Kubeconfig.Clusters = []v1alpha1.ClusterKubeconfigClusterItem{
+			{
+				Name: cluster.Name,
+				Cluster: v1alpha1.ClusterKubeconfigCluster{
+					Server:                   clusterCfg.Server,
+					CertificateAuthorityData: clusterCfg.CertificateAuthorityData,
+				},
+			}}
+	}
+
+	// get oidc info from organization
+	oidc, err := r.getOIDCInfo(ctx, cluster.Namespace)
+	if err != nil {
+		kubeconfig.Status.Conditions.SetConditions(v1alpha1.TrueCondition(v1alpha1.KubeconfigReconcileFailedCondition, "OIDCFetch", err.Error()))
+		updateRequired = true
+		failed = true
+		l.Error(err, "unable to fetch oidc data")
+	} else {
+		oidcHash, err := rxhash.HashStruct(oidc)
+		if err != nil {
+			l.Error(err, "unable to hash oidc info")
+			return ctrl.Result{}, err
+		}
+		if kubeconfig.ObjectMeta.Annotations[OIDCHashAnnotation] != oidcHash {
+			updateRequired = true
+			if kubeconfig.ObjectMeta.Annotations == nil {
+				kubeconfig.ObjectMeta.Annotations = make(map[string]string)
+			}
+			kubeconfig.ObjectMeta.Annotations[OIDCHashAnnotation] = oidcHash
+
+			kubeconfig.Spec.Kubeconfig.AuthInfo = []v1alpha1.ClusterKubeconfigAuthInfoItem{
+				{
+					Name: "oidc@" + cluster.Name,
+					AuthInfo: v1alpha1.ClusterKubeconfigAuthInfo{
+						AuthProvider: clientcmdapi.AuthProviderConfig{
+							Name: "oidc",
+							Config: map[string]string{
+								"client-id":      oidc.ClientID,
+								"client-secret":  oidc.ClientSecret,
+								"idp-issuer-url": oidc.IssuerURL,
+							},
+						},
+					},
+				},
+			}
+		}
+	}
+
+	if updateRequired {
+		if !failed {
+			kubeconfig.Status.Conditions.SetConditions(v1alpha1.TrueCondition(v1alpha1.KubeconfigReadyCondition, "Complete", ""))
+		}
+		// check for the context settings
+		kubeconfig.Spec.Kubeconfig.Contexts = []v1alpha1.ClusterKubeconfigContextItem{
+			{
+				Name: cluster.Name,
+				Context: v1alpha1.ClusterKubeconfigContext{
+					Cluster:   cluster.Name,
+					AuthInfo:  "oidc@" + cluster.Name,
+					Namespace: "default",
+				},
+			}}
+		kubeconfig.Spec.Kubeconfig.CurrentContext = cluster.Name
+
+		err := r.Client.Update(ctx, &kubeconfig)
+		if err != nil {
+			l.Error(err, "unable to update kubeconfig")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+type OIDCInfo struct {
+	ClientID     string
+	ClientSecret string
+	IssuerURL    string
+}
+
+func (r *KubeconfigReconciler) getOIDCInfo(ctx context.Context, orgName string) (OIDCInfo, error) {
+	l := log.FromContext(ctx).WithValues("org", orgName)
+
+	var org v1alpha1.Organization
+	if err := r.Get(ctx, client.ObjectKey{Name: orgName}, &org); err != nil {
+		l.Error(err, "unable to fetch organization", "organization", orgName)
+		return OIDCInfo{}, err
+	}
+
+	clientIDRef := org.Spec.Authentication.OIDCConfig.ClientIDReference
```

Comment:
The OIDCConfig on the Organization CRD is optional and requires to be nil checked. The call to `getOIDCInfo(..)` could be done at the beginning of the reconciliation loop and exit early w.o. error if there is not config set.

Reply:
fixed by be487f5b79f636355d12fc0b34864c545328392a